src/types.rs

```diff
@@ -0,0 +1,51 @@
+//! Platform dependent types.
+
+cfg_if! {
+    if #[cfg(feature = "std")] {
+        pub use std::os::raw::*;
+        use std::borrow::Cow;
+        use std::fmt;
+        use std::path::PathBuf;
+    } else if #[cfg(all(windows, feature = "winapi"))] {
+        pub use winapi::ctypes::*;
+    } else {
+        pub use libc::*;
+    }
+}
+
+/// A platform independent representation of a string. When working with `std`
+/// enabled it is recommended to the convenience methods for providing
+/// conversions to `std` types.
+#[derive(Debug)]
+pub enum BytesOrWideString<'a> {
+    /// A slice, typically provided on Unix platforms.
+    Bytes(&'a [u8]),
+    /// Wide strings typically from Windows.
+    Wide(&'a [u16]),
+}
+
+#[cfg(feature = "std")]
+impl<'a> BytesOrWideString<'a> {
+    /// Lossy converts to a `Cow<str>`, will allocate if `Bytes` is not valid
+    /// UTF-8 or if `BytesOrWideString` is `Wide`.
+    pub fn as_str_lossy(&self) -> Cow<'a, str> {
+        use self::BytesOrWideString::*;
+
+        match self {
+            Bytes(slice) => String::from_utf8_lossy(slice),
+            Wide(wide) => Cow::Owned(String::from_utf16_lossy(wide)),
+        }
+    }
+
+    /// Provides a `Path` representation of `BytesOrWideString`.
+    pub fn into_path_buf(self) -> PathBuf {
+        PathBuf::from(&*self.as_str_lossy())
```

Comment:
This should be an os-specific implementation using the extension traits on Windows while assering that `Wide` doesn't happen on Unix (to ensure this is a lossless conversion)